src/coreclr/vm/jitinterface.cpp

```diff
@@ -11981,6 +11998,42 @@ bool CEEInfo::getReadonlyStaticFieldValue(CORINFO_FIELD_HANDLE fieldHnd, uint8_t
     return result;
 }
 
+bool CEEInfo::readObject(CORINFO_OBJECT_HANDLE handle, uint8_t* buffer, int bufferSize, int valueOffset)
+{
+    CONTRACTL {
+        THROWS;
+        GC_TRIGGERS;
+        MODE_PREEMPTIVE;
+    } CONTRACTL_END;
+
+    _ASSERT(handle != NULL);
+    _ASSERT(buffer != NULL);
+    _ASSERT(bufferSize > 0);
+    _ASSERT(valueOffset >= 0);
+
+    bool result = false;
+
+    JIT_TO_EE_TRANSITION();
+
+    GCX_COOP();
+    OBJECTREF objRef = getObjectFromJitHandle(handle);
+    _ASSERTE(objRef != NULL);
+
+    // Require that the object is immutable, we can relax this later if needed for partially immutable objects
```

Comment:
There is nothing wrong with the JIT reading mutable object if it uses the information correctly (e.g. as a guard).

Reply:
> There is nothing wrong with the JIT reading mutable object if it uses the information correctly (e.g. as a guard).

Ah, I should revert the change and move this check back to JIT side since JIT should not fold accesses to mutubale fields in frozen objects on NativeAOT too

Reply:
@jkotas does VM side look good otherwise (reverted the change)

Reply:
The API looks very similar to `getReadonlyStaticFieldValue`, the difference is in how they specify "this".
- Should `readObject` wrap object references as `CORINFO_OBJECT_HANDLE`, same as `getReadonlyStaticFieldValue`? 
- Can the names be unified (`read` vs. `get`) to make it more obvious that these APIs are very similar?

(I understand that you do not need these for current PR. I am ok if you say that these are TODOs.)

Reply:
> Should readObject wrap object references as CORINFO_OBJECT_HANDLE, same as getReadonlyStaticFieldValue?

Not sure I understand this part - isn't already wrapped? the signature is:
```c
bool readObject(CORINFO_OBJECT_HANDLE handle, uint8_t* buffer, int bufferSize, int valueOffset)
```

> Can the names be unified (read vs. get) to make it more obvious that these APIs are very similar?

Yeah I can rename, to, say `getObjectData` I guess

Reply:
Or you mean when we take an object and read its GC field? I assume that will never happen now but possible in futute

Reply:
> Or you mean when we take an object and read its GC field? I assume that will never happen now but possible in futute

Yes, this is what I meant.